// @ionic-enterprise/identity-vault/dist/esm/BrowserVault.js
import { __rest } from "tslib";
import { VaultErrorCodes } from "./definitions";
const STORAGE_KEYS = {
    DATA: "data",
};
/**
 * THIS VAULT DOES NOT IMPLEMENT SECURE STORAGE IN THE BROWSER. It only exists
 * as a way to run browser-compatible code in place of Identity Vault. Browsers
 * do not have a secure storage element same as native devices. This class
 * is intended to be used to enable running your application in the browser while
 * simulating the functions of Identity Vault using sessionStorage.
 *
 * Represents a vault implementation for browser compatibility.
 */
export class BrowserVault {
    /**
     * @usage
     * ```typescript
     * const vault = new Vault({
     *  key: 'com.company.myvaultapp',
     *  type: 'CustomPasscode',
     *  deviceSecurityType: 'Both',
     *  lockAfterBackgrounded: 2000,
     * });
     * ```
     * @param config
     */
    constructor(config) {
        /** @ignore */
        this.isVaultLocked = false;
        console.warn("THIS VAULT DOES NOT IMPLEMENT SECURE STORAGE IN THE BROWSER AND IS NOT INTENDED FOR PRODUCTION USE. It only exists as a way to run browser-compatible code in place of Identity Vault. Browsers do not have a secure storage element same as native devices. This class is intended to be used to enable running your application in the browser while simulating the functions of Identity Vault using sessionStorage.");
        this.config = Object.assign({
            deviceSecurityType: "Both",
            androidBiometricsPreferStrongVaultOrSystemPasscode: "StrongVault",
            shouldClearVaultAfterTooManyFailedAttempts: false,
            customPasscodeInvalidUnlockAttempts: 5,
            unlockVaultOnLoad: false,
        }, config);
        this.isVaultLocked = config.unlockVaultOnLoad ? false : true;
    }
    /** See {@link Vault.doesVaultExist} */
    doesVaultExist() {
        const data = this.getDataObj();
        return Promise.resolve(!!data);
    }
    /** See {@link Vault.clear} */
    clear() {
        this.unlockIfLocked();
        sessionStorage.removeItem(this.getKey(STORAGE_KEYS.DATA));
        return Promise.resolve();
    }
    /** See {@link Vault.exportVault} */
    exportVault() {
        this.unlockIfLocked();
        const data = this.getDataObj();
        return Promise.resolve(data !== null && data !== void 0 ? data : {});
    }
    /** See {@link Vault.importVault} */
    importVault(data) {
        this.setDataObj(data);
        return Promise.resolve();
    }
    /** See {@link Vault.isLocked} */
    isLocked() {
        return Promise.resolve(this.isVaultLocked);
    }
    /** See {@link Vault.getKeys} */
    getKeys() {
        this.unlockIfLocked();
        const data = this.getDataObj();
        if (!data)
            return Promise.resolve([]);
        return Promise.resolve(Object.keys(data));
    }
    /** See {@link Vault.getValue} */
    getValue(key) {
        var _a;
        this.unlockIfLocked();
        const data = this.getDataObj();
        return Promise.resolve((_a = data === null || data === void 0 ? void 0 : data[key]) !== null && _a !== void 0 ? _a : null);
    }
    /** See {@link Vault.lock} */
    lock() {
        var _a;
        (_a = this.lockCallback) === null || _a === void 0 ? void 0 : _a.call(this);
        return Promise.resolve();
    }
    /** See {@link Vault.removeValue} */
    removeValue(key) {
        this.unlockIfLocked();
        const data = this.getDataObj();
        if (!data)
            return Promise.resolve();
        const _a = data, _b = key, removed = _a[_b], dataAfterRemoval = __rest(_a, [typeof _b === "symbol" ? _b : _b + ""]);
        this.setDataObj(dataAfterRemoval);
        return Promise.resolve();
    }
    /** See {@link Vault.setCustomPasscode} */
    setCustomPasscode(passcode) {
        return Promise.resolve();
    }
    /** See {@link Vault.setValue} */
    setValue(key, value) {
        this.unlockIfLocked();
        const data = this.getDataObj();
        if (!data) {
            this.setDataObj({ [key]: value });
        }
        else {
            this.setDataObj(Object.assign(Object.assign({}, data), { [key]: value }));
        }
        return Promise.resolve();
    }
    /** See {@link Vault.onConfigChanged} */
    onConfigChanged(callback) {
        this.configCallback = callback;
    }
    /** See {@link Vault.onError} */
    onError(callback) {
        this.errorCallback = callback;
    }
    /** See {@link Vault.onLock} */
    onLock(callback) {
        this.lockCallback = callback;
    }
    /** See {@link Vault.onPasscodeRequested} */
    onPasscodeRequested(callback) {
        // No passcode support
    }
    /** See {@link Vault.onUnlock} */
    onUnlock(callback) {
        this.unlockCallback = callback;
    }
    /** See {@link Vault.unlock} */
    unlock() {
        var _a;
        (_a = this.unlockCallback) === null || _a === void 0 ? void 0 : _a.call(this);
        return Promise.resolve();
    }
    /** See {@link Vault.updateConfig} */
    updateConfig(config) {
        var _a;
        this.config = config;
        (_a = this.configCallback) === null || _a === void 0 ? void 0 : _a.call(this, config);
        return Promise.resolve();
    }
    /** @ignore */
    requestBiometricPrompt() {
        this.unlockIfLocked();
        return Promise.resolve(true);
    }
    /** @ignore */
    unlockIfLocked() {
        if (this.isVaultLocked) {
            this.unlock();
        }
    }
    /** @ignore */
    getDataObj() {
        var _a;
        const value = sessionStorage.getItem(this.getKey(STORAGE_KEYS.DATA));
        if (!value)
            return null;
        try {
            return JSON.parse(value);
        }
        catch (e) {
            (_a = this.errorCallback) === null || _a === void 0 ? void 0 : _a.call(this, {
                message: "Unable to parse data store",
                code: VaultErrorCodes.Unknown,
            });
            return null;
        }
    }
    /** @ignore */
    setDataObj(data) {
        var _a;
        try {
            const dataStr = JSON.stringify(data);
            sessionStorage.setItem(this.getKey(STORAGE_KEYS.DATA), dataStr);
        }
        catch (e) {
            (_a = this.errorCallback) === null || _a === void 0 ? void 0 : _a.call(this, {
                message: "Unable to serialize data",
                code: VaultErrorCodes.Unknown,
            });
        }
    }
    /** @ignore */
    getKey(key) {
        return `IV-${this.config.key}-${key}`;
    }
}
